tests/utils_tests/test_text.py

```diff
@@ -95,17 +95,72 @@ def test_truncate_chars(self):
         )
 
     def test_truncate_chars_html(self):
-        perf_test_values = [
-            (("</a" + "\t" * 50000) + "//>", None),
-            ("&" * 50000, "&" * 9 + "…"),
-            ("_X<<<<<<<<<<<>", None),
-        ]
-        for value, expected in perf_test_values:
-            with self.subTest(value=value):
-                truncator = text.Truncator(value)
-                self.assertEqual(
-                    expected if expected else value, truncator.chars(10, html=True)
-                )
+        truncator = text.Truncator(
+            '<p id="par"><strong><em>The quick brown fox jumped over the lazy dog.</em>'
+            "</strong></p>"
+        )
+        self.assertEqual(
+            '<p id="par"><strong><em>The quick brown fox jumped over the lazy dog.</em>'
+            "</strong></p>",
+            truncator.chars(80, html=True),
+        )
+        self.assertEqual(
+            '<p id="par"><strong><em>The quick…</em></strong></p>',
+            truncator.chars(10, html=True),
+        )
+        self.assertEqual(
+            '<p id="par"><strong><em>The qu....</em></strong></p>',
+            truncator.chars(10, "....", html=True),
+        )
+        self.assertEqual(
+            '<p id="par"><strong><em>The quick </em></strong></p>',
+            truncator.chars(10, "", html=True),
+        )
+
+        # Test with new line inside tag
+        truncator = text.Truncator(
+            '<p>The quick <a href="xyz.html"\n id="mylink">brown fox</a> jumped over '
+            "the lazy dog.</p>"
+        )
+        self.assertEqual(
+            '<p>The quick <a href="xyz.html"\n id="mylink">brow…</a></p>',
+            truncator.chars(15, html=True),
+        )
+        self.assertEqual(
+            "<p>Th…</p>",
+            truncator.chars(3, html=True),
+        )
+
+        # Test self-closing tags
+        truncator = text.Truncator(
+            "<br/>The <hr />quick brown fox jumped over the lazy dog."
+        )
+        self.assertEqual("<br/>The <hr />quick brown…", truncator.chars(16, html=True))
+        truncator = text.Truncator(
+            "<br>The <hr/>quick <em>brown fox</em> jumped over the lazy dog."
+        )
+        self.assertEqual(
+            "<br>The <hr/>quick <em>brown…</em>", truncator.chars(16, html=True)
+        )
+
+        # Test html entities
+        truncator = text.Truncator(
+            "<i>Buenos d&iacute;as! &#x00bf;C&oacute;mo est&aacute;?</i>"
+        )
+        self.assertEqual(
+            "<i>Buenos días! ¿Cómo está?</i>",
```

Comment:
So currently we get `í` in place of `&iacute;`. 

It's possible to convert back to HTML elements (?) with `encode('ascii', 'xmlcharrefreplace')` but that gives me back `&#237;`. 🤔 Likely need to decide if this approach is even worthwhile before trying to answer this.

Reply:
I think in the context of truncating í is more useful than `&iacute;` since the client will only show one char and also because truncating inside an entity is made impossible by converting entities to characters.

I wouldn't advocate for removing entities everywhere but in the context of displaying a truncated result to users it should be fine. 

(Converting back to entities came up a few times; a person who wanted it found a partial solution here using BeautifulSoup https://github.com/matthiask/html-sanitizer/issues/23 )

Reply:
I'm glad to provide you with more information on the change made in Django's codebase!

In Django, the Truncator class in the django.utils.text module is used to truncate text. Prior to the change made in ticket #16421, Truncator used regular expressions to parse and truncate HTML content. However, regular expressions are not a reliable way to parse HTML, and may lead to unexpected results or errors.

As a result, in ticket #30686, the Truncator class was updated to use Python's built-in HTMLParser module instead of regular expressions. This change made the Truncator class more reliable and accurate when handling HTML content.

Using the HTMLParser module, Truncator can now properly handle nested HTML tags and attributes, and avoid accidentally cutting off tags or attributes in the middle. The HTMLParser module is a more robust and flexible way to handle HTML parsing, making Truncator more reliable and accurate when handling HTML content.

Overall, this change improved the accuracy and reliability of the Truncator class when handling HTML content, which is an important feature for many Django applications.